### ISSUE TITLE
Training on low res images

### ISSUE THREAD

**Issue by JonathanCMitchell** [matterport/Mask_RCNN#321]
Hi there, what would I need to change in the config / loss function in order to train this on lower resolution images? I would like to train on (256, 256, 3) images instead of (1024, 1024, 3) images. When I changed
```
IMAGE_MIN_DIM = 800 //4
IMAGE_MAX_DIM=800//4
USE_MINI_MASK=False
```
This seemed to work, but I started to get nan's for my rpn loss value. Any suggestions on what I might need to change in order to train on lower resolution images? (I am hoping that just by changing the IMAGE_MAX_DIM the code will actually resize my images for me). If that's not the case please let me know.

Thanks,
Edited 03/13/18 ===
```
   RES_SCALE_FACTOR = 4
    RPN_ANCHOR_SCALES = (32//RES_SCALE_FACTOR, 64//RES_SCALE_FACTOR, 128//RES_SCALE_FACTOR, 256//RES_SCALE_FACTOR, 512//RES_SCALE_FACTOR)
    IMAGE_MIN_DIM = 800 //RES_SCALE_FACTOR
    IMAGE_MAX_DIM = 1024 //RES_SCALE_FACTOR
    USE_MINI_MASK=False
```

**Comment by FruVirus** [comment]
@JonathanCMitchell 

If all of your training images have the same shape, then there is no need to resize the images. In load_image_gt(), you can bypass the calls to utils.resize_image() and utils.resize_mask().

**Comment by JonathanCMitchell** [comment]
@FruVirus I am trying to train on the COCO dataset, but downscale the resolution to (256, 256, 3)

**Comment by FruVirus** [comment]
@JonathanCMitchell 

Ah ok, then in that case, you might be running into some issues due to the way image resizing is done. You'll likely have to change utils.resize_image() so that it downscales both dimensions of the COCO image to the same size.

In utils.resize_image(), you want the call to scipy.misc.imresize() to output the same dimensions for both height and width (i.e., 256 x 256). The way that the function is currently written, unless the original COCO image is square, this won't be the case since the same scale factor is used for both height and width.

A simple change would be to calculate a different scale factor for height and width and pass those to the scipy.misc.imresize() call (also to the mask resizing).

However, if you had IMAGE_PADDING = True in your config, the resized image should still be square so it's likely that the NANs for the RPN loss is caused by something else (e.g., higher learning rate, no gradient clipping perhaps?).

**Comment by JonathanCMitchell** [comment]
I also think I may have to change the anchor box sizes in the config right? They are given in terms of pixel values so I would have to divide all of them by 4 if I wanted to train on (256, 256, 3). @FruVirus

**Comment by FruVirus** [comment]
@JonathanCMitchell 

Yes, good call. Especially since a 512 x 512 anchor makes no sense if your images are 256 x 256 lol

**Comment by JonathanCMitchell** [comment]
Yeah @FruVirus I think the nan's might be due to gradient clipping or something. It doesn't seem to be a sizing issue, although the Nan comes from the rpn_bbox_loss, so that may be an indicator of sorts.

**Comment by FruVirus** [comment]
@JonathanCMitchell 

There's a thread here that talks about nan values from rpn_bbox_loss. I'm not sure what the thread is exactly but it shouldn't be hard to find---maybe that addresses your issue?

**Comment by JonathanCMitchell** [comment]
@FruVirus I saw [215](https://github.com/matterport/Mask_RCNN/issues/215) that discusses training on low res images and [159](https://github.com/matterport/Mask_RCNN/issues/159) that seems to contain information regarding a recent pull request that may fix the `NaN` errors I was experiencing.

**Comment by Cpruce** [comment]
@JonathanCMitchell Did you try with coco image shapes for min and max to see if there's any other problem?

**Comment by waleedka** [comment]
The PR fix for the NaN issue was merged.

@JonathanCMitchell To answer the original question about using low resolution imagse:

1. Scaling the image is correct. And, disabling mini masks is not a bad option as well. 
2. Anchors: Scaling them like you did is a good option. Another option is to keep the same scales but remove anchors from higher layers that exceed the 256x256 size.
3. You might also want to be aware of the code in `PyramidROIAlign` that assigns ROIs to different layers depending on their area. Since you're using smaller images, it's likely that your P5 and P6 layers might not get assigned any ROIs anyway. Not that it's an issue, just be aware of it.

**Comment by JonathanCMitchell** [comment]
@Cpruce 
I changed 
```
IMAGE_MIN_DIM = 200
IMAGE_MAX_DIM = 256
```
And I kept padding=True so it should be a square (256, 256, 3) RGB Image. I verified that this is what was getting passed in as training data. Although I do not know if I should be scaling something inside the data generator. Have you trained on a different resolution size? 

@waleedka 
For (2) How would I remove anchors that exceed 256, 256 in size? 
For (3) Thank you for this heads up! That is probably the case. 

I should also note that I am using a MobileNet architecture similar to the one presented in @Cpruce 's [repo](https://github.com/Cpruce/Mask_RCNN)

**Comment by JonathanCMitchell** [comment]
@FruVirus If I keep padding = True, then it seems to be resizing from (170, 256, 3) to (256, 256, 3) but adds a bunch of zeros for pad on the first dimension. I believe this may be causing me to get NaN's when I actually divide the values inside `np.log(gt_h / a_h)` inside build_rpn_targets.

**Comment by FruVirus** [comment]
@JonathanCMitchell 

Yea, that makes sense and good to know. For images with dimensions smaller than 256, those dimensions will be zero-padded.

Perhaps what you want is anisotropic resizing to 256 x 256 without zero-padding similar to what they did in the original R-CNN paper.

**Comment by FruVirus** [comment]
@JonathanCMitchell 

As a first approach, can you try just resizing each image to 256 x 256 regardless of its original dimensions and without zero-padding?

This would mean that you need to modify utils.resize_image(). In particular, this line:

image = scipy.misc.imresize(image, (round(h * scale), round(w * scale)))

"scale" for height and width would be different if the original image is not square to begin with.

**Comment by JonathanCMitchell** [comment]
@FruVirus Yes, that would work for the image, but I would need a similar way of resizing the mask, because the mask must be assuming that I am padding.

**Comment by waleedka** [comment]
> it seems to be resizing from (170, 256, 3) to (256, 256, 3) but adds a bunch of zeros for pad on the first dimension. I believe this may be causing me to get NaN's when I actually divide the values inside np.log(gt_h / a_h) inside build_rpn_targets.

Can you elaborate on this? The padding shouldn't cause an issue, unless there is a bug we're missing.

**Comment by JonathanCMitchell** [comment]
@waleedka  would (3) still matter if I am using the mini mask?

**Comment by waleedka** [comment]
@JonathanCMitchell by (3), I think you're referring to my point about how ROIs get assigned to different layers. That shouldn't have anything to do with mini-masks. 

I'm guessing the NaN issue is fixed, so I'm closing this. If the NaN still happens, please provide more details (preferably in a separate issue).

**Comment by ApoorvaSuresh** [comment]
Hi,

how was the NaN issue fixed finally?
